File: HW8/BigramMap.java
/**
 *
 * @author devo35
 * @param <T>
 *
 * Dosya Ismi: BigramMap.java
 * Yazan: Devrim AKINCI
 * Numarasi: 141044052
 * Tarih: 06 Ocak 2017
 */

import java.io.File;
import java.io.FileNotFoundException;
import java.util.InputMismatchException;
import java.util.Scanner;
import java.util.NoSuchElementException;

public class BigramMap<T> implements Bigram<T>{
    private int size;//Dosyadaki eleman sayisi
    private int [] freq;//Tekrar eden ikilileri tutar
    private Object [] data;//Verileri tutan degisken
    private int dataType;//Veri tipi
    private Scanner input;//Dosyayi okumak icin
    private Object [] first;
    private Object [] second;
    
    public BigramMap(int type){//Constructor
        dataType = type;
        size = 0;
    }
    
    public void readFile(String fileName){
        String temp;
        try{
            input = new Scanner(new File(fileName));//Dosyanin acilmasi
/*---------------------Dosyadaki eleman sayisinin bulunmasi-------------------*/                        
            do{
                try{
                    temp = input.next();
                    size++;
                }                
                catch(NoSuchElementException inputException){
                    //Dosyanin ilk satiri bos ise program exception firlatir.
                    System.err.println("First line of file is empty!");
                    throw inputException;
                }
            }while(input.hasNext());
        }
        catch (FileNotFoundException fileNotFound){
            System.err.println("Error:This file could not be opened!");
        }
/*----------------------------------------------------------------------------*/        
        input.close();//Dosyanin kapatilmasi
        try{
            input = new Scanner(new File(fileName));//Dosyanin acilmasi
/*-------------------------Hata kontrolu--------------------------------------*/                        
            if (dataType == 1){
//Dosyanin icinde double veya string tipi bir veri varsa program exception firlatir.                                
                try{
                    data = new Integer[size];
                    for (int i=0; i<data.length; i++){   
                        data[i] = input.nextInt();//Verilerin okunmasi
                    }
                }
                catch(InputMismatchException inputException){
                    System.err.println("Invalid file. There are string or double data types in file!");
                    throw inputException;
                }
            }
            else if (dataType == 2){
                try{
                    data = new String[size];
                    for (int i=0; i<data.length; i++){
                        data[i] = input.next();//Verilerin okunmasi
                    }
                }
                catch(InputMismatchException inputException){
                    System.err.println("Invalid file!");
                    throw inputException;       
                }
            }
            else if (dataType == 3){
//Dosyanin icinde integer veya string tipi bir veri varsa program exception firlatir.                                                
                try{
                    String tmp;
                    tmp = input.nextLine();//Verilerin okunmasi
                    data = new Double[size];                
                    for (int i=0; i<data.length; i++){
                        data[i] = Double.parseDouble(tmp.split(" ")[i]);                        
                    }
                }
                catch(NumberFormatException inputException){
                    System.err.println("Invalid file. There are integer or string data types in file!");
                    throw inputException;
                }
            }
        }
        catch (FileNotFoundException fileNotException){
            System.err.println("Error:This file could not be opened!");
        } 
/*----------------------------------------------------------------------------*/        
    }
    public int numGrams(){
        return size - 1;
    }
    public int numOfGrams(Object firstElement, Object secondElement){
        int count = 0;//Verilen ikilinin veri dizisi icinde kac tane oldugunu tutar.
        if (((dataType == 1) || (dataType == 2)) || (dataType == 3)){
            for (int i=0; i<size; i++){
                if ((data[i].equals(firstElement)) && (data[i+1].equals(secondElement)))
                    count++;
            }
        }
        return count;
    }
    public String toString(){
        int temp;
        Object temp1,temp2;
        freq = new int[size-1];
        if (dataType == 1){
            first = new Integer[size-1];
            second = new Integer[size-1];
        }
        else if (dataType == 2){
            first = new String[size-1];
            second = new String[size-1];
        }
        else if (dataType == 3){
            first = new Double[size-1];
            second = new Double[size-1];
        }
        for (int i=0; i<size-1; i++){
            freq[i] = numOfGrams(data[i], data[i+1]);
            first[i] = data[i];
            second[i] = data[i+1];
        }
        for (int i=0; i<size-1; i++){
            for (int j=i+1; j<size-1; j++){
        //Tekrar eden ikilinin sayilarinin buyukten kucuge dogru siralanmasi                                            
                if (freq[i]<freq[j]){
                    //Swap islemi
                    temp=freq[i];
                    temp1=first[i];
                    temp2=second[i];
                    freq[i]=freq[j];
                    first[i]=first[j];
                    second[i]=second[j];
                    freq[j]=temp;
                    first[j]=temp1;
                    second[j]=temp2;
                }
            }
        }
        //Yazdirma islemi
        if (dataType == 1){
            for (int i=0; i<size-2; i++)
                System.out.printf("(%d,%d)-> %d\n",first[i],second[i],freq[i]);
            return String.format("(%d,%d)-> %d\n",first[size-2],second[size-2],freq[size-2]);
        }
        else if (dataType == 2){
            for (int i=0; i<size-2; i++)
            System.out.printf("(%s,%s)-> %d\n",first[i],second[i],freq[i]);
        return String.format("(%s,%s)-> %d\n",first[size-2],second[size-2],freq[size-2]);
        }
        else if (dataType == 3){
            for (int i=0; i<size-2; i++)
            System.out.printf("(%.1f,%.1f)-> %d\n",first[i],second[i],freq[i]);
        return String.format("(%.1f,%.1f)-> %d\n",first[size-2],second[size-2],freq[size-2]);
        }
        return String.format("");
    }
}

File: HW7/BigramMap.cpp
/* 
 * Dosya Ismi: BigramMap.cpp
 * Yazan: Devrim AKINCI
 * Numarasi: 141044052
 * Tarih: 21 Aralik 2016
 */

#include "BigramMap.h"

template <class T>
BigramMap<T>::BigramMap(int type){
    dataType = type;
}

template <class T>
BigramMap<T>::~BigramMap() {
    delete [] data;
}

template <class T>
void BigramMap<T>::readFile(string fileName){
    fstream file;
    string temp;
    data = new T[size];
    file.open(fileName);//Dosyanin acilmasi
    if(file.is_open()){
/*---------------------Dosyadaki eleman sayisinin bulunmasi-------------------*/        
        file >> temp;
        if (temp == "")
            throw myException(-1);
        do{
/*-------------------------Hata kontrolu--------------------------------------*/            
            if (dataType == 1){
//Dosyanin icinde double veya string tipi bir veri varsa program exception firlatir.                
            if(temp[0] < '0' || temp[0] > '9')
                throw myException(dataType);
            else if (searchDot(temp,'.') == 1)
                throw myException(dataType);
            }
            else if (dataType == 2){
//Dosyanin icinde double veya integer tipi bir veri varsa program exception firlatir.                
                if ('0' <= temp[0] && temp[0] <= '9')
                    throw myException(dataType);
            }
            else if (dataType == 3){
//Dosyanin icinde integer veya string tipi bir veri varsa program exception firlatir.                                
                if(temp[0] < '0' || temp[0] > '9')
                    throw myException(dataType);
                else if (searchDot(temp,'.') == 0)
                    throw myException(dataType);
            }
            else//Dosyanin ilk satiri bos ise program exception firlatir.
                throw myException(0);
/*----------------------------------------------------------------------------*/            
            size++;  
        }while(file >> temp);
/*----------------------------------------------------------------------------*/        
    }
    else//Eger dosya hatali ise program exception firlatir.
    	throw myException();
    file.close();//Dosyanin kapanmasi
    data = new T[size];
    file.open(fileName);
    if (file.is_open()){//Dosyanin acilmasi
    	for (int i=0; i<size; i++)
            file >>data[i];//Verilerin okunmasi
    }
    else//Eger dosya hatali ise program exception firlatir.
    	throw myException();
    file.close();//Dosyanin kapanmasi
}
template <class T>
int BigramMap<T>::numGrams(){
    return size - 1;
}
template <class T>
int BigramMap<T>::numOfGrams(T firstElement,T secondElement){
    int count=0;//Verilen ikilinin veri dizisi icinde kac tane oldugunu tutar
    if (((dataType == 1) || (dataType == 2)) || (dataType == 3)){
    	for (int i=0; i<size; i++){
            if ((data[i] == firstElement) && (data[i+1] == secondElement))
                count++;
    	}
    }
    return count;
}
template <class T>
pair <T,T>BigramMap<T>::maxGrams(){
    pair <T,T> maxPair;
    setPair (data);
    for (int i=0; i<size-1; i++)
        myMap[i] = numOfGrams(data[i],data[i+1]);
    int max = myMap[0];
    maxPair.first = data[0];
    maxPair.second = data[1];
    for (int i=1; i<size-1; i++){//En cok tekrar eden ikilinin bulunmasi
        if (myMap[i] > max){
            myMap[i] = numOfGrams(data[i],data[i+1]);   
            maxPair.first = data[i];
            maxPair.second = data[i+1];
        }
    }
    return maxPair;
}
template <class T>
int BigramMap<T>::searchDot(const string str, char target) const{
    int result=0;//Dosyanin icinde nokta karakterinin sayisini tutar
    for (int i=0; str[i] != '\0'; i++){
        if (str[i] == target)//Nokta karakterinin bulunmasi
            result++;
    }
    return result;
}

template <class T>
ostream& BigramMap<T>::print(ostream& out){
    int temp;
    pair <T,T>tempPair;
    for (int i=0; i<size-1; i++){
        for (int j=i+1; j<size-1; j++){
//Tekrar eden ikilinin sayilarinin buyukten kucuge dogru siralanmasi            
            if (myMap[i]<myMap[j]){
                //Swap islemi
            	temp=myMap[i];
                myMap[i]=myMap[j];
                myMap[j]=temp;
                tempPair = myPair[i];
                myPair[i] = myPair[j];
                myPair[j] = tempPair;
            }
        }
    }
    for (int i=0; i<size-1; i++)
        cout << "("<< myPair[i].first << "," <<myPair[i].second << ")->"
        << myMap[i] << endl;
    return out;
}
template <class T>
void BigramMap<T>::setPair (T* fileData){
	myPair = new pair<T,T>[size-1];
	for (int i=0; i<size-1; i++){
    	myPair[i].first = fileData[i];
        myPair[i].second = fileData[i+1];
	}
}
